src/bun.js/api/bun/socket.zig

```diff
@@ -1910,7 +2082,495 @@ fn NewSocket(comptime ssl: bool) type {
+            // where the standard requirements as described in ISO/IEC 9899:2011
+            // sections 7.21.2.1, 7.21.1.2, and 7.1.4, would be violated.
+            // Thus, we use a dummy byte.
+            var dummy: [1]u8 = undefined;
+            const size = BoringSSL.SSL_get_peer_finished(ssl_ptr, @ptrCast(*anyopaque, &dummy), @sizeOf(@TypeOf(dummy)));
+            if (size == 0) return JSValue.jsUndefined();
+
+            var buffer = bun.default_allocator.alloc(u8, @intCast(usize, size)) catch unreachable;
+            defer bun.default_allocator.free(buffer);
+
+            _ = BoringSSL.SSL_get_peer_finished(ssl_ptr, @ptrCast(*anyopaque, buffer.ptr), buffer.len);
+            return JSC.ArrayBuffer.createBuffer(globalObject, buffer);
+        }
+
+        pub fn getTLSFinishedMessage(
+            this: *This,
+            globalObject: *JSC.JSGlobalObject,
+            _: *JSC.CallFrame,
+        ) callconv(.C) JSValue {
+            if (comptime ssl == false) {
+                return JSValue.jsUndefined();
+            }
+
+            if (this.detached) {
+                return JSValue.jsUndefined();
+            }
+
+            const ssl_ptr = @ptrCast(*BoringSSL.SSL, this.socket.getNativeHandle());
+            // We cannot just pass nullptr to SSL_get_finished()
+            // because it would further be propagated to memcpy(),
+            // where the standard requirements as described in ISO/IEC 9899:2011
+            // sections 7.21.2.1, 7.21.1.2, and 7.1.4, would be violated.
+            // Thus, we use a dummy byte.
+            var dummy: [1]u8 = undefined;
+            const size = BoringSSL.SSL_get_finished(ssl_ptr, @ptrCast(*anyopaque, &dummy), @sizeOf(@TypeOf(dummy)));
+            if (size == 0) return JSValue.jsUndefined();
+
+            var buffer = bun.default_allocator.alloc(u8, @intCast(usize, size)) catch unreachable;
+            defer bun.default_allocator.free(buffer);
+
+            _ = BoringSSL.SSL_get_finished(ssl_ptr, @ptrCast(*anyopaque, buffer.ptr), buffer.len);
+            return JSC.ArrayBuffer.createBuffer(globalObject, buffer);
+        }
+
+        pub fn getSharedSigalgs(
+            this: *This,
+            globalObject: *JSC.JSGlobalObject,
+            _: *JSC.CallFrame,
+        ) callconv(.C) JSValue {
+            JSC.markBinding(@src());
+            if (comptime ssl == false) {
+                return JSValue.jsNull();
+            }
+
+            if (this.detached) {
+                return JSValue.jsNull();
+            }
+            const ssl_ptr = @ptrCast(*BoringSSL.SSL, this.socket.getNativeHandle());
+
+            const nsig = BoringSSL.SSL_get_shared_sigalgs(ssl_ptr, 0, null, null, null, null, null);
+
+            const array = JSC.JSValue.createEmptyArray(globalObject, @intCast(usize, nsig));
+
+            for (0..@intCast(usize, nsig)) |i| {
+                var hash_nid: c_int = 0;
+                var sign_nid: c_int = 0;
```

Comment:
i haven't finished reading the PR yet but i hope there are tests for all this stuff because it will be really hard to know if we broke any of this in like 2 months

Reply:
I wrote tests but `getSharedSigalgs` needs `sigalgs` option so we can do more reliable tests on it.